File: test/document/location_test.rb
# frozen_string_literal: true

require "test_helper"

module Document
  class LocationTest < Minitest::Spec
    test "document.location returns whole document location" do
      document = Yerba::Document.parse(<<~YAML)
        name: Alice
        port: 5432
      YAML

      loc = document.location

      assert_instance_of Yerba::Location, loc
      assert_equal 1, loc.start_line
      assert_equal 0, loc.start_column
    end

    test "document.location with selector returns location of value" do
      document = Yerba::Document.parse(<<~YAML)
        name: Alice
        port: 5432
      YAML

      loc = document.location("port")

      assert_equal 2, loc.start_line
      assert_equal 6, loc.start_column
    end

    test "document.location with nested selector" do
      document = Yerba::Document.parse(<<~YAML)
        database:
          host: localhost
      YAML

      loc = document.location("database.host")

      assert_equal 2, loc.start_line
    end

    test "document.location with bracket selector" do
      document = Yerba::Document.parse(<<~YAML)
        - id: talk-1
          title: First
        - id: talk-2
          title: Second
      YAML

      loc = document.location("[1].title")

      assert_equal 4, loc.start_line
      assert_equal 4, loc.end_line
      assert_equal 9, loc.start_column
      assert_equal 15, loc.end_column
    end

    test "document.location returns nil for missing selector" do
      document = Yerba::Document.parse(<<~YAML)
        name: Alice
      YAML

      assert_nil document.location("nonexistent")
    end

    test "document.location spans multiple lines for block scalars" do
      document = Yerba::Document.parse(<<~YAML)
        - id: talk-1
          description: |-
            First line.
            Second line.
            Third line.
        - id: talk-2
      YAML

      loc = document.location("[0].description")

      assert_equal 2, loc.start_line
      assert_equal 5, loc.end_line
      assert_equal 15, loc.start_column
      assert_equal 15, loc.end_column
    end

    test "document.location for single line value has same start and end line" do
      document = Yerba::Document.parse(<<~YAML)
        name: Alice
      YAML

      loc = document.location("name")

      assert_equal loc.start_line, loc.end_line
    end

    test "document.location for map spans all its keys" do
      document = Yerba::Document.parse(<<~YAML)
        - id: talk-1
          title: First
          description: |-
            Some text.
            More text.
        - id: talk-2
      YAML

      loc = document.location("[0]")

      assert_equal 1, loc.start_line
      assert_equal 5, loc.end_line
    end

    test "document.location has byte offsets" do
      document = Yerba::Document.parse(<<~YAML)
        name: Alice
      YAML

      loc = document.location("name")

      assert_equal 6, loc.start_offset
      assert_equal 11, loc.end_offset
    end

    test "document.locations returns array for wildcard selectors" do
      document = Yerba::Document.parse(<<~YAML)
        - id: a
          title: First
        - id: b
          title: Second
      YAML

      locs = document.locations("[].title")

      assert_equal 2, locs.length
      assert_equal 2, locs[0].start_line
      assert_equal 4, locs[1].start_line
    end

    test "document.locations returns all items" do
      document = Yerba::Document.parse(<<~YAML)
        - id: a
        - id: b
        - id: c
      YAML

      locs = document.locations("[]")

      assert_equal 3, locs.length
      assert_equal 1, locs[0].start_line
      assert_equal 2, locs[1].start_line
      assert_equal 3, locs[2].start_line
    end

    test "document.locations with nested wildcards" do
      document = Yerba::Document.parse(<<~YAML)
        - speakers:
            - Alice
            - Bob
        - speakers:
            - Charlie
      YAML

      locs = document.locations("[].speakers[]")

      assert_equal 3, locs.length
      assert_equal 2, locs[0].start_line
      assert_equal 3, locs[1].start_line
      assert_equal 5, locs[2].start_line
    end

    test "document.locations returns empty array for no matches" do
      document = Yerba::Document.parse(<<~YAML)
        name: Alice
      YAML

      locs = document.locations("[].nonexistent")

      assert_equal [], locs
    end
  end
end

File: rust/tests/location_test.rs
mod support;
use indoc::indoc;
use support::parse;

#[test]
fn test_location_of_scalar() {
  let document = parse(indoc! {"
    name: Alice
    port: 5432
  "});

  let info = document.get_node_info("name");

  assert_eq!(info.location.start_line, 1);
  assert_eq!(info.location.start_column, 6);
  assert_eq!(info.location.end_line, 1);
  assert_eq!(info.location.end_column, 11);
}

#[test]
fn test_location_of_nested_scalar() {
  let document = parse(indoc! {"
    database:
      host: localhost
      port: 5432
  "});

  let info = document.get_node_info("database.host");

  assert_eq!(info.location.start_line, 2);
  assert_eq!(info.location.start_column, 8);
  assert_eq!(info.location.end_line, 2);
  assert_eq!(info.location.end_column, 17);
}

#[test]
fn test_location_of_map() {
  let document = parse(indoc! {"
    database:
      host: localhost
      port: 5432
  "});

  let info = document.get_node_info("database");

  assert_eq!(info.location.start_line, 2);
  assert_eq!(info.location.end_line, 3);
}

#[test]
fn test_location_of_sequence() {
  let document = parse(indoc! {"
    tags:
      - ruby
      - rust
  "});

  let info = document.get_node_info("tags");

  assert_eq!(info.location.start_line, 2);
  assert_eq!(info.location.end_line, 3);
}

#[test]
fn test_location_of_indexed_item() {
  let document = parse(indoc! {"
    - id: talk-1
      title: First
    - id: talk-2
      title: Second
  "});

  let info = document.get_node_info("[1].title");

  assert_eq!(info.location.start_line, 4);
  assert_eq!(info.location.start_column, 9);
  assert_eq!(info.location.end_line, 4);
  assert_eq!(info.location.end_column, 15);
}

#[test]
fn test_location_of_block_scalar() {
  let document = parse(indoc! {"
    - id: talk-1
      description: |-
        First line.
        Second line.
        Third line.
    - id: talk-2
  "});

  let info = document.get_node_info("[0].description");

  assert_eq!(info.location.start_line, 2);
  assert_eq!(info.location.end_line, 5);
}

#[test]
fn test_location_of_root_map_item() {
  let document = parse(indoc! {"
    - id: talk-1
      title: First
      description: |-
        Some text.
        More text.
    - id: talk-2
  "});

  let info = document.get_node_info("[0]");

  assert_eq!(info.location.start_line, 1);
  assert_eq!(info.location.end_line, 5);
}

#[test]
fn test_location_has_byte_offsets() {
  let document = parse(indoc! {"
    name: Alice
  "});

  let info = document.get_node_info("name");

  assert_eq!(info.location.start_offset, 6);
  assert_eq!(info.location.end_offset, 11);
}

#[test]
fn test_location_of_second_line_scalar() {
  let document = parse(indoc! {"
    name: Alice
    port: 5432
  "});

  let info = document.get_node_info("port");

  assert_eq!(info.location.start_line, 2);
  assert_eq!(info.location.start_column, 6);
  assert_eq!(info.location.start_offset, 18);
  assert_eq!(info.location.end_offset, 22);
}

#[test]
fn test_resolve_selectors_for_locations() {
  let document = parse(indoc! {"
    - id: a
      title: First
    - id: b
      title: Second
    - id: c
      title: Third
  "});

  let selectors = document.resolve_selectors("[].title");

  assert_eq!(selectors, vec!["[0].title", "[1].title", "[2].title"]);
}

#[test]
fn test_resolve_selectors_for_nested_locations() {
  let document = parse(indoc! {"
    - speakers:
        - Alice
        - Bob
    - speakers:
        - Charlie
  "});

  let selectors = document.resolve_selectors("[].speakers[]");

  assert_eq!(selectors, vec!["[0].speakers[0]", "[0].speakers[1]", "[1].speakers[0]"]);
}

#[test]
fn test_location_of_each_resolved_selector() {
  let document = parse(indoc! {"
    - id: a
    - id: b
    - id: c
  "});

  let selectors = document.resolve_selectors("[]");

  assert_eq!(selectors.len(), 3);

  let info0 = document.get_node_info(&selectors[0]);
  let info1 = document.get_node_info(&selectors[1]);
  let info2 = document.get_node_info(&selectors[2]);

  assert_eq!(info0.location.start_line, 1);
  assert_eq!(info1.location.start_line, 2);
  assert_eq!(info2.location.start_line, 3);
}
